File: custom-bots/node-gateway/src/infrastructure/tracing.ts
import { diag, DiagConsoleLogger, DiagLogLevel, trace } from "@opentelemetry/api";
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";
import { BatchSpanProcessor } from "@opentelemetry/sdk-trace-base";
import { Resource } from "@opentelemetry/resources";
import { SemanticResourceAttributes } from "@opentelemetry/semantic-conventions";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { registerInstrumentations } from "@opentelemetry/instrumentation";
import { HttpInstrumentation } from "@opentelemetry/instrumentation-http";
import { FastifyInstrumentation } from "@opentelemetry/instrumentation-fastify";
import { PgInstrumentation } from "@opentelemetry/instrumentation-pg";

diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.INFO);

export function setupTracing(): void {
  const exporterEndpoint =
    process.env.OTEL_EXPORTER_OTLP_ENDPOINT ?? "http://otel-collector:4318/v1/traces";

  const resource = new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: "node-gateway",
  });

  const provider = new NodeTracerProvider({ resource });

  const exporter = new OTLPTraceExporter({ url: exporterEndpoint });

  provider.addSpanProcessor(new BatchSpanProcessor(exporter));
  provider.register();

  registerInstrumentations({
    tracerProvider: provider,
    instrumentations: [new HttpInstrumentation(), new FastifyInstrumentation(), new PgInstrumentation()],
  });
}

export function getActiveTraceContext():
  | {
      traceId: string;
      spanId: string;
    }
  | undefined {
  const span = trace.getActiveSpan();
  if (!span) {
    return undefined;
  }
  const ctx = span.spanContext();
  return {
    traceId: ctx.traceId,
    spanId: ctx.spanId,
  };
}

File: custom-bots/biometric-proxy/app/infrastructure/tracing.py
from __future__ import annotations

import os

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.semconv.resource import ResourceAttributes


def setup_tracing() -> None:
    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://otel-collector:4318/v1/traces")

    resource = Resource(
        attributes={
            ResourceAttributes.SERVICE_NAME: "biometric-proxy",
        }
    )

    provider = TracerProvider(resource=resource)
    span_exporter = OTLPSpanExporter(endpoint=endpoint)
    span_processor = BatchSpanProcessor(span_exporter)

    provider.add_span_processor(span_processor)
    trace.set_tracer_provider(provider)


def get_active_trace_context() -> dict[str, str] | None:
    span = trace.get_current_span()
    if not span:
        return None
    ctx = span.get_span_context()
    return {
        "trace_id": ctx.trace_id,
        "span_id": ctx.span_id,
    }

